crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -363,6 +363,32 @@ pub fn derive_system_param(input: TokenStream) -> TokenStream {
     let generics = ast.generics;
     let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
 
+    let lifetime_generics: Vec<_> = generics.lifetimes().map(|l| &l.lifetime).collect();
+
+    let impl_generics_as_generics = {
+        let tmp = TokenStream::from(quote!(#impl_generics));
+        parse_macro_input!(tmp as Generics)
+    };
+
+    let (lifetime_impl_generics, lifetime_generic) = match lifetime_generics.len() {
```

Comment:
Rather than matching on `.len()`, you can match on `.as_slice()` and directly on `[]` and `[lifetime_generic]` for the two first case.
<https://blog.rust-lang.org/inside-rust/2020/03/04/recent-future-pattern-matching-improvements.html#subslice-patterns-head-tail-->

Reply:
Yeah that is definitely better, always forget rust supports this. Fixed!